### ISSUE TITLE
Maybe more than 25%!!!!!!! of the game is a fake spectrum now!!!

### ISSUE THREAD

**Issue by ainixingo** [gcp/leela-zero#1906]
![image](https://user-images.githubusercontent.com/40348917/46399110-71b3de80-c729-11e8-8997-8f0a7d23a56e.png)
![image](https://user-images.githubusercontent.com/40348917/46399124-7d9fa080-c729-11e8-82fe-4f098d453760.png)
![image](https://user-images.githubusercontent.com/40348917/46399141-88f2cc00-c729-11e8-85b5-16294bac3542.png)
![image](https://user-images.githubusercontent.com/40348917/46399161-97d97e80-c729-11e8-8834-88ddcc72ffce.png)
![image](https://user-images.githubusercontent.com/40348917/46399232-c2c3d280-c729-11e8-9eab-7c16f380aded.png)
![image](https://user-images.githubusercontent.com/40348917/46399258-d0795800-c729-11e8-9f0b-a1c14661d352.png)
![image](https://user-images.githubusercontent.com/40348917/46399277-dbcc8380-c729-11e8-866b-c4723b6632a4.png)
![qq 20181003124659](https://user-images.githubusercontent.com/40348917/46399289-e25afb00-c729-11e8-8d1d-cd65a55403c3.png)
![qq 20181003124656](https://user-images.githubusercontent.com/40348917/46399290-e2f39180-c729-11e8-9caf-8b2ed8b46148.png)
![qq 20181003124653](https://user-images.githubusercontent.com/40348917/46399291-e2f39180-c729-11e8-990a-f864d571a706.png)
![qq 20181003124649](https://user-images.githubusercontent.com/40348917/46399292-e2f39180-c729-11e8-9fd7-c2c15fe0a427.png)
![qq 20181003124646](https://user-images.githubusercontent.com/40348917/46399293-e38c2800-c729-11e8-8fa7-1ee4dadb998d.png)
![qq 20181003124641](https://user-images.githubusercontent.com/40348917/46399294-e38c2800-c729-11e8-8fc8-a28f81ab6403.png)
![qq 20181003124638](https://user-images.githubusercontent.com/40348917/46399295-e38c2800-c729-11e8-9b58-e5b35241ec47.png)
![qq 20181003124633](https://user-images.githubusercontent.com/40348917/46399297-e424be80-c729-11e8-873c-1af05dc03550.png)
![qq 20181003124625](https://user-images.githubusercontent.com/40348917/46399299-e424be80-c729-11e8-94b9-d50a22bdd43a.png)
![qq 20181003124711](https://user-images.githubusercontent.com/40348917/46399300-e424be80-c729-11e8-9517-20ca34d3fdd6.png)
![qq 20181003124708](https://user-images.githubusercontent.com/40348917/46399302-e4bd5500-c729-11e8-978c-82feee4f2951.png)
![qq 20181003124705](https://user-images.githubusercontent.com/40348917/46399304-e4bd5500-c729-11e8-93bf-e33162f3f9a0.png)
![qq 20181003124702](https://user-images.githubusercontent.com/40348917/46399305-e555eb80-c729-11e8-9b98-6b193f197d9d.png)

**Comment by seopsx** [comment]
I can confirm that it is the same client number I found today uploading malicious match games.

https://github.com/gcp/leela-zero/issues/458#issuecomment-426521756

See #458 and #1705

**Comment by l1t1** [comment]
bad behavior

**Comment by wonderingabout** [comment]
as leela zero project grows bigger and have more people interested in it, it's natural that more troublemakers are coming too

leela zero will have to enforce its security filtering of games at some point, or the situation will only get worse

**Comment by ainixingo** [comment]
![image](https://user-images.githubusercontent.com/40348917/46403618-9910a880-c735-11e8-87a2-fb900d5b8e59.png)
![image](https://user-images.githubusercontent.com/40348917/46403626-9e6df300-c735-11e8-9fdd-f9f83ea524b2.png)
![image](https://user-images.githubusercontent.com/40348917/46403643-aa59b500-c735-11e8-84f1-c0120acded46.png)
![image](https://user-images.githubusercontent.com/40348917/46403661-b5144a00-c735-11e8-93e2-0e5f0e7ecb8b.png)
![image](https://user-images.githubusercontent.com/40348917/46403677-c4939300-c735-11e8-9371-ce806d296fb0.png)
![image](https://user-images.githubusercontent.com/40348917/46403691-d2e1af00-c735-11e8-9e5f-044c64203df4.png)
![image](https://user-images.githubusercontent.com/40348917/46403705-dbd28080-c735-11e8-82c7-3fbdfa6ce6c7.png)

**Comment by ainixingo** [comment]
Maybe more than 30%!!!!!!! of the game is a fake spectrum now!!! 30%  @gcp

**Comment by ainixingo** [comment]
This is no longer a urine at sea. It is a serious problem and must be resolved as soon as possible. @gcp

**Comment by herazul** [comment]
Relax man, GCP and Roy will handle the situation when they will have time to check github.

**Comment by Friday9i** [comment]
By default, clients uploading several games per minute should probably be cross-checked! We'd be happy to get users contributing with huge hardware, but as of now, this is suspect...

**Comment by ihavnoid** [comment]
Training games intentionally have randomness, occasionally having silly moves. Can you elaborate what makes it look wrong?

**Comment by PhilipFRipper** [comment]
Some of them look like normal training hijinx to me.  But there are a few examples in there that are suspect, making several unlikely moves that are not the kind of random move LZ tends to self play into.

Like the first game shown.  That's very unlikely.  Each of those moves taken along could be all the noise, but the odds of them coming in sequence are very low.

**Comment by seopsx** [comment]
> Training games intentionally have randomness, occasionally having silly moves. Can you elaborate what makes it look wrong?

That client should be banned and all of its games removed based on its match behaviour alone. It's most likely the same client "6" as seen here today https://github.com/gcp/leela-zero/issues/458#issuecomment-426521756 
And if the games come from Google, the malevolent behaviour of IP address should be reported to them so they can sort it out.

**Comment by Hayden2018** [comment]
This is a very serious issue and may affect the strength of leela.

**Comment by Ishinoshita** [comment]
In this respect, I would be curious to analyse the policy distribution of LZ181/hash 68824bbc training data, to see if it's standard or not (with supposedly 25% of the games, I would expect something should be visible) and to review such games anotated with policy values (like in this kind of sgf that I cooked for LZ177 [chunk_1042_game_7.zip](https://github.com/Ishinoshita/Leela-Zero-Training-Data-Viewer/files/2435779/chunk_1042_game_7.zip)).

But,unfortunately, LZ181 training data are not yet available in the [Index of /training/](https://leela.online-go.com/training/?C=M&O=D). Not even LZ180's yet. Is there a way to speed-up the release of some chunks (say a few 100's) ?

**Comment by dvdkhlng** [comment]
I noticed leela zero playing strange moves during training which I attributed to noise (maybe you call this temperature).  If some games cannot be explained by noise, maybe they can be explained by software bugs.  OpenCL bugs may look like that.  E.g. I've seen my PC acting stangely when suspended in the middle of a leela zero self-play run, so I can imagine a lot to go wrong in OpenCL (just look at the amount of code in mesa's opencl implementation, how would this ever be bug-free?).

There may be a simple test to (occasionally) check for broken (or malicuous) clients (or at least estimate ratio of broken clients): make clients play a game without any noise, maybe at a low playout setting, then check the results against move-probabilities computed server-side.  At a low noise setting, most client games will be similar (when using the same network), so probable variations can be pre-computed and/or cashed, so that the effort of checking results will grow sub-linearly with the number of nodes.

**Comment by seopsx** [comment]
@qzq1 
One way to judge would be to look at the match games played by the same client.
I don't know the IP (maybe someone could check?) but this is the same client number "6" from the same morning:

 http://zero.sjeng.org/viewmatch/43cba0db9b6de6e26655aec74466d882ccf10f5927b8c3e5dd2c9306300c3a60?viewer=wgo

Bad moves: C2, P16 (big drop in winrate)

http://zero.sjeng.org/viewmatch/09d5c9a22754eee22ac1902a855722094bb6f8b7a4f8bc740af134adb5d6fccb?viewer=wgo

Bad moves: E4, G7, F6, R3, P17

I didn't find a single correct match game. Here are some more evidently malicious:

http://zero.sjeng.org/viewmatch/4a249a8b31c531c9eb599a2dc82727ecf173fcf9d1d67e5f5f08be4ed45d9f7c?viewer=wgo

http://zero.sjeng.org/viewmatch/4f766abe8726435e04784efe7d9641aed6959a3a10b02df7a391c165580d74fe?viewer=wgo

http://zero.sjeng.org/viewmatch/cb1ab7ccacba38d3a78443d1140a838f85941c5d1a42326438eca9446a83d826?viewer=wgo

![bad move examples1](https://user-images.githubusercontent.com/35587223/46461913-9d39d600-c7c8-11e8-9206-bdceeefe29de.png)
![bad_games5](https://user-images.githubusercontent.com/35587223/46461933-ae82e280-c7c8-11e8-87a2-a94ecb7f206f.png)

**Comment by ainixingo** [comment]
![image](https://user-images.githubusercontent.com/40348917/46470290-b2cef000-c808-11e8-9748-f0c7b5aad616.png)
![image](https://user-images.githubusercontent.com/40348917/46470329-cd08ce00-c808-11e8-92e1-02d87db42bf6.png)
![image](https://user-images.githubusercontent.com/40348917/46470400-0e00e280-c809-11e8-9fa6-20d571e393d7.png)
![image](https://user-images.githubusercontent.com/40348917/46470428-22dd7600-c809-11e8-89e9-ae88342e68fd.png)
![image](https://user-images.githubusercontent.com/40348917/46470452-3f79ae00-c809-11e8-9643-233b73c66860.png)

**Comment by gjm11** [comment]
@seops @ainixingo Have you tried the following experiment? Run Sabaki or some such UI.. Play LZ against itself _using the same parameters used for self-play games_ a few times, and see whether you see a similar rate of obviously-bad-looking moves. I think this means `-v 1600 -m 30 -n`.

Maybe others can do better, but so far all I'm getting from this issue report is that sometimes some self-play games have bad-looking moves in them. That's not news, and it's not a problem; as I'm sure you know, self-play games are deliberately randomized a bit. If you think there's _more than that_ going on then let's have some explanation of why ordinary self-play randomness isn't enough to explain what you're seeing.

A dozen pictures each showing one dubious-looking move don't tell us anything. You could at least explain, e.g., whether they're all from different games or several from the same one. Or why the moves look too bad to be explained by ordinary self-play randomness.

**Comment by PhilipFRipper** [comment]
I think explaining the difference between some of these games and normal self play requires a level of go knowledge that only one subset of the board has.  The other subset has code / ai knowledge but limited go knowledge and probably can't tell the difference, which seems pretty obvious to go players.

I'm not sure how I could explain in layman's terms, except to say that some of these are much worse than I've seen in other self play games.  To the point where it "feels" like someone else.  Maybe a much less trained network - like someone trying to train their own and not turning off the GTP stuff.

**Comment by seopsx** [comment]
@PhilipFripper @Ishinoshita 
You can find Leela Zero options and seed from the SGF comment section and test if you can reproduce the game. Just replace "?viewer=wgo" in the URL with .sgf, download and  open the sgf-file with a text editor.

**Comment by alreadydone** [comment]
@wonderingabout This looks like usual behavior under low winrate + randomness. If it were not a no-resign game it would probably be resigned at move 92/93. Can anyone check?

**Comment by dvdkhlng** [comment]
Did anybody else look into the **match** games linked by @qzq1 above?  They really do look wrong.  And match games are played without noise, so can't be explained by that.  For example [this](http://zero.sjeng.org/viewmatch/4a249a8b31c531c9eb599a2dc82727ecf173fcf9d1d67e5f5f08be4ed45d9f7c?viewer=wgo) game.  
Also the kind of strange play is too non-random to be explained by noise.

Just a few bad match-games would easily tip the scales of a match, as those are usually quite tight.  What would be the impact on Leela Zero's learning process if the wrong "best" network were selected?

**Comment by Splee99** [comment]
> What would be the impact on Leela Zero's learning process if the wrong "best" network were selected?

I would think the result is that another new best network will be easily generated.

**Comment by foreverpokemongo** [comment]
> > What would be the impact on Leela Zero's learning process if the wrong "best" network were selected?
> 
> I would think the result is that another new best network will be easily generated.

If the attack persists (if exists at all), then the so-called 'new best network' is not the best. The actual strength may stall or degrade.

**Comment by wonderingabout** [comment]
@alreadydone it's true that until 49 all moves are reasonable, but then white also plays a lot of 2nd/1st line moves too, like 64 84 88 90 and so on
it may just be high winrate behaviour too, i dont know

**Comment by Ishinoshita** [comment]
@wonderingabout It is both high and low win rates, as they alternate when the game is  completely out of balance, that causes the search to flatten the raw policy and open the door to temperature picking weird moves with high frequency.

And there are many games that would be resigned very early, except for the minimum 92 move count. I clearly remember many games of LZ156 that were lost cause at move 30, and of series of 1st line moves or the like as soon as move 40. The distribution of game length show a peak at 92 moves, artifact of this situation.

**Comment by qzq1** [comment]
@Ishinoshita that makes perfect sense, and would seem to explain the mixture of weird-looking and normal-looking games.  Do you have a sense for what that means for learning?  Does LZ effectively disregard moves after the point that the game is effectively decided?  Or does the phenomenon demonstrate a need to improve learning by incorporating dynamic komi or some other method to promote consistently strong moves?
